### PR TITLE
Introduce task_runner_checker, task_runner_weak_ptr

### PR DESCRIPTION
**Pull request by cyanglaz**
Replace thread_checker with task_runner_checker, make the weakPtr to be accessed from different threads as long as they are on the same task_runner (thread merging)
https://github.com/flutter/flutter/issues/52253

### CODE REVIEW

File: fml/memory/task_runner_checker.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      return true;
+    }
+    if (queues->Owns(self_, current_queue_id)) {
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  TaskQueueId self_;
```
Comment (iskakaushik) [inline fml/memory/task_runner_checker.h:42]: I think `self_` is an ambiguous name. Can this be called, `initialized_queue_id_`?

File: fml/memory/task_runner_checker.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+namespace fml {
+
+class TaskRunnerChecker final {
+ public:
+  TaskRunnerChecker() : self_(InitTaskQueueId()){};
+
+  ~TaskRunnerChecker() = default;
+
+  bool RunsOnCreationTaskRunner() const {
+    FML_DCHECK(fml::MessageLoop::IsInitializedForCurrentThread());
```
Comment (iskakaushik) [inline fml/memory/task_runner_checker.h:23]: I think this can be a `FML_CHECK`.

File: fml/memory/task_runner_checker.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      return true;
+    }
+
+    auto queues = MessageLoopTaskQueues::GetInstance();
+    if (queues->Owns(current_queue_id, self_)) {
+      return true;
+    }
+    if (queues->Owns(self_, current_queue_id)) {
+      return true;
+    }
```
Comment (iskakaushik) [inline fml/memory/task_runner_checker.h:37]: Can you refactor to share code with `TaskRunner::RunsTasksOnCurrentThread`?

File: fml/memory/task_runner_checker.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ private:
+  TaskQueueId self_;
+
+  TaskQueueId InitTaskQueueId() {
+    MessageLoop::EnsureInitializedForCurrentThread();
+    return MessageLoop::GetCurrentTaskQueueId();
+  }
+};
+
+#if !defined(NDEBUG)
```
Comment (iskakaushik) [inline fml/memory/task_runner_checker.h:50]: shouldn't this condition be flipped? As it stands it looks like these checks will only be run when `NDEBUG` is not specified. I think we should rely on `FLUTTER_RUNTIME_MODE` macro instead and run it on `DEBUG` builds.

**Comment by cyanglaz** [comment]
@googlebot I fixed it.

File: fml/memory/weak_ptr.h [hunk 5]
```diff
@@ -43,7 +48,7 @@ class WeakPtr {
   WeakPtr() : ptr_(nullptr) {}
 
   // Copy constructor.
-  WeakPtr(const WeakPtr<T>& r) = default;
+  explicit WeakPtr(const WeakPtr<T>& r) = default;
```
Comment (cyanglaz) [inline fml/memory/weak_ptr.h:51]: @iskakaushik Changed after your last approval:
I found implicit casting could be problematic as someone could want a TaskRunnerAffineWeakPtr but end up getting a WeakPtr.

File: shell/common/shell.h [hunk 6]
```diff
@@ -378,8 +378,9 @@ class Shell final : public PlatformView::Delegate,
   std::unique_ptr<ShellIOManager> io_manager_;   // on IO task runner
   std::shared_ptr<fml::SyncSwitch> is_gpu_disabled_sync_switch_;
 
-  fml::WeakPtr<Engine> weak_engine_;          // to be shared across threads
-  fml::WeakPtr<Rasterizer> weak_rasterizer_;  // to be shared across threads
+  fml::WeakPtr<Engine> weak_engine_;  // to be shared across threads
+  fml::TaskRunnerAffineWeakPtr<Rasterizer>
```
Comment (cyanglaz) [inline shell/common/shell.h:382]: @iskakaushik Changed after your last approval:
It was implicitly casted from a `TaskRunnerAffineWeakPtr` to a `WeakPtr`

File: fml/memory/weak_ptr.h (hunk 5 above)
Comment (iskakaushik) [reply fml/memory/weak_ptr.h:51]: LGTM

**Pull request merged**